### DISCUSSION TITLE
Best option to print labels

### DISCUSSION THREAD

**Discussion by HannesMorre** [inventree/InvenTree#5994] [Q&A]
I am wondering what the best option is for me to print labels of all stock items. Currently we have around 1300 stock items and I would like to print labels with qr codes on them for all these stock items. We have a Brother QL-500, Dymo LabelWriter 450 Duo and Zebra ZT2300.
What would be our best option to efficiently print labels for all these and future stock items? I cannot get the allignment of these labels right when downloading them and then printing on one of these printers.

I know there are plugins for zebra, dymo and brother, but we cannot manage to get them installed.
If someone could help with this, that would be great.

**Comment by wolflu05** [comment]
The best option would be to connect a label printer via a plugin to InvenTree. That way you can print labels from any InvenTree client (web, app) and you are not limited to a computer where the special label printer software runs on.

1. First you have to create your own label template if you want to not use the default ones with only the QR code/(name).
2. Next you would need to setup the printing plugin by following the documentation provided by the plugin. You said you don't get that to work. Please provide some more detail of what you did, what doesn't work and what errors/logs you see, otherwise we cannot help you
3. If the printing plugin works, you can go to the stock items table, select all stock items and choose the print label action. After selecting the template and printer, this will print labels for all of your stock items with just a few clicks.

**Comment by HannesMorre** [comment]
Okay, I somehow did manage to install the brother plugin, now how do I use it? I cannot find any configuration documentation.
![image](https://github.com/inventree/InvenTree/assets/71696745/9b02292d-aa81-49ef-9e55-efd0ec50a238)

**Reply by wolflu05** [reply]
You need to configure it first before you can print with it. To do so click on the name of the plugin in the settings page on the left navigation, below the plugins page.

**Reply by HannesMorre** [reply]
I suppose it should be here somewhere?
![image](https://github.com/inventree/InvenTree/assets/71696745/354d3099-c677-46bb-a097-475dc6402e28)
How can I make it visible in this navigation bar?

**Reply by miggland** [reply]
Yes, it should be in that list. However, it's a bit weird that your plugin doesn't list the Version and author (it does on my installation). Have you restarted the server after installing and activating the plugin?

**Reply by HannesMorre** [reply]
Not yet, I will ask our system admin to reset the server and see what happens. I don't know if this makes any difference, but I installed the plugin with the "Install Plugin" button you can see on the screenshot. Our system admin tried to install the plugins directly from the commandline on the server, but this did not work.

**Reply by HannesMorre** [reply]
We just restarted the server and nothing changed.

**Reply by SchrodingersGat** [reply]
@HannesMorre I assume you have also done a refresh on the settings page?

1. What version of InvenTree are you running?
2. If you navigate to the [admin interface](https://docs.inventree.org/en/stable/settings/admin/) and navigate to "plugins / plugin configurations" - do you see an entry for the brother label plugin?
3. When you install (via the web interface) what does the "success" message say?

**Reply by HannesMorre** [reply]
Yes I did refresh the settings page.
1. # Version Information:
InvenTree-Version: 0.12.7
Django Version: 3.2.19
Commit Hash: 6563b4c
Commit Date: 2023-09-16

Database: postgresql
Debug-Mode: False
Deployed using Docker: True
Platform: Linux-5.10.0-25-amd64-x86_64-with-glibc2.36
Installer: DOC

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}]
2. 
![image](https://github.com/inventree/InvenTree/assets/71696745/acb2540b-14ce-4ddd-9be7-e68077f6209d)
3. 
![image](https://github.com/inventree/InvenTree/assets/71696745/252ff050-dec8-4ebb-a611-18f5fc7a3bed)

**Reply by SchrodingersGat** [reply]
As you have installed via docker, can you share the contents of `plugins.txt` in your external data volume?

**Reply by HannesMorre** [reply]
```
root@1e430943aea8:/home/inventree/data# cat plugins.txt 
# InvenTree Plugins (uses PIP framework to install)
inventree-brother-plugin
git+https://github.com/wolflu05/inventree-cups-plugin
inventree-brother-plugin  # Installed 2023-10-06 13:50:29.312354+00:00 by root
inventree-brother-plugin  # Installed 2023-10-06 13:51:36.282868+00:00 by root
inventree-brother-plugin  # Installed 2023-11-27 15:45:18.710889+00:00 by hmo
inventree-brother-plugin  # Installed 2023-11-28 09:40:48.287806+00:00 by hmo
```

**Reply by SchrodingersGat** [reply]
This might be a similar issue [to this one](https://github.com/inventree/inventree-brother-plugin/issues/25). I have not been able to reproduce the issue myself, however.

**Reply by HannesMorre** [reply]
Hmm okay, weird that I'm not the only one with this issue.

**Comment by HannesMorre** [comment]
FYI

We also tried to install the cups plugin for dymo labelwriters and then we got this error:
```
root@1e430943aea8:/home/inventree/InvenTree# pip install git+https://github.com/wolflu05/inventree-cups-plugin
Collecting git+https://github.com/wolflu05/inventree-cups-plugin
  Cloning https://github.com/wolflu05/inventree-cups-plugin to /tmp/pip-req-build-1z0yzjs8
  Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-cups-plugin /tmp/pip-req-build-1z0yzjs8
  Resolved https://github.com/wolflu05/inventree-cups-plugin to commit e22e825507b6355d9d35cd286ebff0d13b79ef88
  Preparing metadata (setup.py) ... done
Collecting pycups (from inventree-cups-plugin==0.1.0)
  Downloading pycups-2.0.1.tar.gz (62 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.0/63.0 kB 2.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: inventree-cups-plugin, pycups
  Building wheel for inventree-cups-plugin (setup.py) ... done
  Created wheel for inventree-cups-plugin: filename=inventree_cups_plugin-0.1.0-py3-none-any.whl size=4838 sha256=608f38cc4c462acc9e712e04278992055e13c3b8c20512cf327681640e620da0
  Stored in directory: /tmp/pip-ephem-wheel-cache-a4rheck8/wheels/66/63/be/3b7871e0fdd5376a3391cb05760a576c062f1ea841a5d5a9c5
  Building wheel for pycups (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      In file included from cupsconnection.c:22:
      cupsconnection.h:26:10: fatal error: cups/http.h: No such file or directory
         26 | #include <cups/http.h>
            |          ^~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pycups
  Running setup.py clean for pycups
Successfully built inventree-cups-plugin
Failed to build pycups
ERROR: Could not build wheels for pycups, which is required to install pyproject.toml-based projects
```

**Reply by SchrodingersGat** [reply]
For the cups printer, you will need the cups binaries installed

**Reply by wolflu05** [reply]
Exactly, there is a little warning note in the [Readme](https://github.com/wolflu05/inventree-cups-plugin?tab=readme-ov-file#installation):
> ⚠️ This plugin needs cups-devel installed to install its dependencies. You can read more about the requirements at pycups. If you're using apt as a package manager run `apt install libcups2-dev` before.

If you're using docker, you can either create your own image based on the Inventree image or run the apt command for testing directly in the exceced shell, but note that this doesn't persist during container recreation.

**Reply by HannesMorre** [reply]
Our admin apparently already tried this before and ran into another error then.
He will try it again and we will post the results here to keep you posted.

Thanks for the help already.

**Comment by SchrodingersGat** [comment]
@HannesMorre FYI I have just created a fresh docker install using image `inventree:0.12.7` and then run the following commands:

### Docker Container Setup

```bash
docker compose run inventree-server invoke update
docker compose run inventree-server invoke setup-test -I
docker compose up -d
```

### Plugin Install

Installed `inventree-brother-plugin` (via the web intterface). Returned success message.

### Reboot Docker Image

```bash
docker compose restart inventree-server
```

### Verify Plugin Installation

Confirmed that the plugin config is now visible in the settings interface:

<img width="1277" alt="Screenshot 2023-11-28 at 11 11 46 pm" src="https://github.com/inventree/InvenTree/assets/10080325/ed7ebe87-f01c-4716-a63a-56ec1a1579f3">

Given that this is the exact same docker image you are using, it is very strange that you are seeing different results...

**Comment by lqdmist** [comment]
Hi, I am a co-worker of @HannesMorre.

The plugin has been installed successfully since these two plugins (inventree-cups-plugin & inventree-brother-plugin) are present and enabled in the UI.
I created a Dockerfile where I installed the `libcups2-dev` package as suggested by @wolflu05.
I also installed the two plugins using the Dockerfile:

``` Dockerfile
FROM inventree/inventree:0.12.10

## Dependency for the cups plugin: https://inventree.org/wolflu05/inventree-cups-plugin
RUN apt-get update && apt install libcups2-dev -y

## Install plugins:
RUN pip install git+https://github.com/wolflu05/inventree-cups-plugin
RUN pip install inventree-brother-plugin
```

Whenever I try to print out a labels using either one of the two plugins, I stumble on the same issue:

``` bash
inventree-worker    | Plugin registry has no record of plugin 'brother'
inventree-worker    | Could not find matching plugin for 'brother'
```
and
``` bash
inventree-worker    | Plugin registry has no record of plugin 'cups'
inventree-worker    | Could not find matching plugin for 'cups'
```
Printer side everything works fine since we are able to print some test labels.

We noticed this issue on our staging environment, so I have setup another instance (0.12.10) from scratch using docker and stumbled on the exact same issue.
Any suggestions?

Edit:
I just tried to install the plugins only using the UI instead of using the Dockerfile `pip install` but same result

**Reply by SchrodingersGat** [reply]
- Do you see the same error on the `inventree-server` container or just the `inventree-worker` container?
- Are `inventree-server` and `inventree-worker` running the exact same image?

Also, you should shell into the docker container, and explore the python directory to see if the `inventree-brother-plugin` plugin has actually been installed correctly.

**Reply by lqdmist** [reply]
These log lines are only written by `inventree-worker`.
Both the `inventree-worker` and `inventree-server` are on the same version which is 0.12.10.

The plugins are present:
```
root@cf290bd7fec1:/usr/local/lib/python3.9/site-packages# ls -al | grep inventree
drwxr-xr-x  2 root root   4096 Dec 13 08:38 brother_ql_inventree-1.0.dist-info
drwxr-xr-x  3 root root   4096 Dec 13 08:38 inventree_brother
drwxr-xr-x  2 root root   4096 Dec 13 08:38 inventree_brother_plugin-0.8.0.dist-info
drwxr-xr-x  3 root root   4096 Dec 13 08:38 inventree_cups
drwxr-xr-x  2 root root   4096 Dec 13 08:38 inventree_cups_plugin-0.1.0.dist-info
root@cf290bd7fec1:/usr/local/lib/python3.9/site-packages# ls -al inventree_brother inventree_cups
inventree_brother:
total 28
-rw-r--r-- 1 root root    0 Dec 13 08:38 __init__.py
drwxr-xr-x 2 root root 4096 Dec 13 08:38 __pycache__
-rw-r--r-- 1 root root 6964 Dec 13 08:38 brother_plugin.py
-rw-r--r-- 1 root root   75 Dec 13 08:38 version.py

inventree_cups:
total 24
-rw-r--r-- 1 root root    0 Dec 13 08:38 __init__.py
drwxr-xr-x 2 root root 4096 Dec 13 08:38 __pycache__
-rw-r--r-- 1 root root 3705 Dec 13 08:38 cups_plugin.py
-rw-r--r-- 1 root root   72 Dec 13 08:38 version.py

```

**Reply by SchrodingersGat** [reply]
When listing the plugins above, were you in the `inventree-worker` container or the `inventree-server` container?

**Reply by lqdmist** [reply]
Those were from the `inventree-server`. I then installed them in the `inventree-worker` container as well but I got the same results.

**Reply by lqdmist** [reply]
Any suggestions? We would really appreciate if these plugins can work since we need them a lot for our production line.

**Reply by matmair** [reply]
@lqdmist have you updated to 0.13.0 or are you still on 0.12.7?

**Reply by lqdmist** [reply]
I initially tried on 0.12.7 but without luck, I then have set up a new clean environment using docker with 0.13.0 but had the same results

**Reply by matmair** [reply]
I have a hard time reproducing - have you tried if you can reproduce this on another system?

**Reply by lqdmist** [reply]
Yes I have, I tried this on Debian 11 using Docker Compose v2.21.0

**Reply by wolflu05** [reply]
Did you got it working?

**Reply by lqdmist** [reply]
No sadly we were unable to get it working

**Reply by wolflu05** [reply]
I really think you didn't used your custom image also for the worker container. Please double check that. While trying to reproduce this, I also added some docker [installation notes](https://github.com/wolflu05/inventree-cups-plugin#docker) to my repository. But unfortunately I cannot reproduce your issue.
